src/hotspot/share/gc/shenandoah/shenandoahGenerationalEvacuationTask.cpp

```diff
@@ -0,0 +1,288 @@
+
+
+  {
+    ShenandoahHeapLocker locker(_heap->lock());
+
+    HeapWord* update_watermark = region->get_update_watermark();
+
+    // Now that this region is affiliated with old, we can allow it to receive allocations, though it may not be in the
+    // is_collector_free range.
+    region->restore_top_before_promote();
+
+    size_t region_used = region->used();
+
+    // The update_watermark was likely established while we had the artificially high value of top.  Make it sane now.
+    assert(update_watermark >= region->top(), "original top cannot exceed preserved update_watermark");
+    region->set_update_watermark(region->top());
+
+    // Unconditionally transfer one region from young to old. This represents the newly promoted region.
+    // This expands old and shrinks new by the size of one region.  Strictly, we do not "need" to expand old
+    // if there are already enough unaffiliated regions in old to account for this newly promoted region.
+    // However, if we do not transfer the capacities, we end up reducing the amount of memory that would have
+    // otherwise been available to hold old evacuations, because old available is max_capacity - used and now
+    // we would be trading a fully empty region for a partially used region.
+    young_gen->decrease_used(region_used);
+    young_gen->decrement_affiliated_region_count();
+
+    // transfer_to_old() increases capacity of old and decreases capacity of young
+    _heap->generation_sizer()->force_transfer_to_old(1);
+    region->set_affiliation(OLD_GENERATION);
+
+    old_gen->increment_affiliated_region_count();
+    old_gen->increase_used(region_used);
+
+    // add_old_collector_free_region() increases promoted_reserve() if available space exceeds plab_min_size()
+    _heap->free_set()->add_promoted_in_place_region_to_old_collector(region);
+  }
+}
+
+void ShenandoahGenerationalEvacuationTask::promote_humongous(ShenandoahHeapRegion* region) {
+  ShenandoahMarkingContext* marking_context = _heap->marking_context();
+  oop obj = cast_to_oop(region->bottom());
+  assert(_heap->gc_generation()->is_mark_complete(), "sanity");
+  shenandoah_assert_generations_reconciled();
+  assert(region->is_young(), "Only young regions can be promoted");
+  assert(region->is_humongous_start(), "Should not promote humongous continuation in isolation");
+  assert(region->age() >= _tenuring_threshold, "Only promote regions that are sufficiently aged");
+  assert(marking_context->is_marked(obj), "promoted humongous object should be alive");
+
+  // TODO: Consider not promoting humongous objects that represent primitive arrays.  Leaving a primitive array
+  // (obj->is_typeArray()) in young-gen is harmless because these objects are never relocated and they are not
+  // scanned.  Leaving primitive arrays in young-gen memory allows their memory to be reclaimed more quickly when
+  // it becomes garbage.  Better to not make this change until sizes of young-gen and old-gen are completely
```

Comment:
So... are sizes of young and old 'completely adaptive' now, and would this optimization make sense?

Reply:
I agree.  I believe there are no longer any blockers on this optimization.

Reply:
@earthling-amzn will spin this out into a JBS ticket on his PR to clean up TODOs.

Reply:
Tracking this under https://bugs.openjdk.org/browse/JDK-8339612